lib/galaxy/tools/toolbox/base.py

```diff
@@ -79,7 +79,13 @@ def __init__( self, config_filenames, tool_root_dir, app, tool_conf_watcher=None
         if tool_conf_watcher:
             self._tool_conf_watcher = tool_conf_watcher  # Avoids (re-)starting threads in uwsgi
         else:
-            self._tool_conf_watcher = get_tool_conf_watcher(lambda: self.handle_reload_toolbox())
+            if hasattr(self.app, 'tool_cache'):
+                # Normal galaxy instances should have a tool_cache,
+                # but the toolshed does not.
+                tool_cache = self.app.tool_cache
+            else:
+                tool_cache = None
+            self._tool_conf_watcher = get_tool_conf_watcher(reload_callback=lambda: self.handle_reload_toolbox(), tool_cache=tool_cache)
```

Comment:
That was my first instinct, but I was thinking that it may be better to have just a single thread watching for toolbox reloads. If we trigger a toolbox reload because the shed_tool_conf and the tool xml have changed, we reload just once now, but if we have separate threads doing the same thing we would need to synchronize them somehow.